UW 537 & UW 538: Add docs for MPAS and MPAS Init drivers. (#450)
ufs-community/uwtools: docs/sections/user_guide/yaml/components/mpas.rst

.. _mpas_yaml:

mpas
====

Structured YAML to run the WRF preprocessing component ``mpas`` is validated by JSON Schema and requires the ``mpas:`` block, described below. If ``mpas`` is to be run via a batch system, the ``platform:`` block, described :ref:`here <platform_yaml>`, is also required.

Here is a prototype UW YAML ``mpas:`` block, explained in detail below:

.. highlight:: yaml
.. literalinclude:: ../../../../shared/mpas.yaml

> When MPAS is built from source code, static data input files and default namelist and streams files are also created. Here we show an example of how an arbitrarily named ``user:`` section may be used to define helpful variables that are heavily reused through the rest of the configuration. In this example, the ``mpas_app:`` entry is the path to where the MPAS model was installed.

Author: More feedback on this wording would be great!

UW YAML for the ``mpas:`` Block
-------------------------------

execution:
^^^^^^^^^^

See :ref:`here <execution_yaml>` for details.

boundary_conditions:
^^^^^^^^^^^^^^^^^^^^

Describes the boundary condition files needed for the forecast. These will be the output from the
``init_atmosphere`` executable, which may be run using the ``mpas_init`` UW driver. Please see its
documentation :ref:`here <mpas_init_yaml>`.

interval_hours:
"""""""""""""""

Frequency interval of the given files, in integer hours.

offset:
"""""""

How many hours earlier the external model used for boundary conditions started compared to the desired forecast cycle, in integer hours.

path:
"""""

An absolute path to the MPAS-ready files to be used initial and lateral boundary condition input. The names of the files are specified in the MPAS streams.atmosphere XML file, and may be specified in the ``streams: values:`` block of the driver YAML.

> 

Reviewer: "used _**for**_ initial and lateral boundary"

length:
^^^^^^^

The length of the forecast in integer hours.

files_to_copy:
^^^^^^^^^^^^^^

See :ref:`this page <files_yaml>` for details.

files_to_link:
^^^^^^^^^^^^^^

Identical to ``files_to_copy:`` except that symbolic links will be created in the run directory instead of copies.

run_dir:
^^^^^^^^

The path to the directory where ``mpas`` will find its namelist, streams file, and necessary data
files and write its outputs.

streams:
^^^^^^^^

path:
"""""

The path to the base ``streams.atmosphere`` file that comes from the MPAS build.

values:
"""""""

> The set of key-value pairs that will render the appropriate XML entries in the streams input file.

Author: I purposefully kept the `streams:` section vague to grow with our XML transition, which we discussed in our backlog grooming session today.

.. _mpas_yaml:

mpas
====

> Structured YAML to run the WRF preprocessing component ``mpas`` is validated by JSON Schema and requires the ``mpas:`` block, described below. If ``mpas`` is to be run via a batch system, the ``platform:`` block, described :ref:`here <platform_yaml>`, is also required.

Reviewer 2: ```suggestion
Structured YAML to run MPAS is validated by JSON Schema and requires the ``mpas:`` block, described below. If ``mpas`` is to be run via a batch system, the ``platform:`` block, described :ref:`here <platform_yaml>`, is also required.
```

Author: Oops!

Here is a prototype UW YAML ``mpas:`` block, explained in detail below:

.. highlight:: yaml
.. literalinclude:: ../../../../shared/mpas.yaml

> When MPAS is built from source code, static data input files and default namelist and streams files are also created. Here we show an example of how an arbitrarily named ``user:`` section may be used to define helpful variables that are heavily reused through the rest of the configuration. In this example, the ``mpas_app:`` entry is the path to where the MPAS model was installed.

Reviewer 2: ```suggestion
An MPAS build provides prototype versions of certain required runtime files. Here, an arbitrarily named ``user:`` block defines an ``mpas_app`` variable, pointing to the directory where MPAS was installed, to reduce duplication in references to those files.
```

``init_atmosphere`` executable, which may be run using the ``mpas_init`` UW driver. Please see its
documentation :ref:`here <mpas_init_yaml>`.

interval_hours:
"""""""""""""""

Frequency interval of the given files, in integer hours.

offset:
"""""""

How many hours earlier the external model used for boundary conditions started compared to the desired forecast cycle, in integer hours.

path:
"""""

An absolute path to the MPAS-ready files to be used for initial and lateral boundary conditions input. The names of the files are specified in the MPAS ``streams.atmosphere`` XML file, and may be specified in the ``streams: values:`` block of the driver YAML.

length:
^^^^^^^

The length of the forecast in integer hours.

files_to_copy:
^^^^^^^^^^^^^^

See :ref:`this page <files_yaml>` for details.

To reduce duplication of information in this section, it may be helpful to template the file that depends on the number of cores being used to run the executable. For example, instead of:

.. code-block:: text

   mpas:
     files_to_copy:
       conus.graph.info.part.32: /path/to/conus.graph.info.part.32

> Jinja expressions can be used to reference the number of cores used in execution:

Reviewer 2: ```suggestion
Jinja2 expressions can be used to reference the number of cores used in execution:
```
Annoying, I now. The Jinja2 people have a GitHub Issue where they discussed trying to change the name officially from Jinja2 to just Jinja and decided not to.

.. code-block:: text

   mpas:
     files_to_copy:
       conus.graph.info.part.{{mpas.execution["batchargs"]["cores"]}}: /path/to/conus.graph.info.part.{{mpas.execution["batchargs"]["cores"]}}

> Note that the ``files_to_copy:`` and the ``file_to_link:`` blocks are the only ones that accept Jinja expressions in their keys.

Reviewer 2: ```suggestion
Note that the ``files_to_copy:`` and the ``file_to_link:`` blocks are the only ones that accept Jinja2 expressions in their keys.
```

> Note that the ``files_to_copy:`` and the ``file_to_link:`` blocks are the only ones that accept Jinja expressions in their keys.

Reviewer 2: Is this true? I think `dereference()` handles Jinja2 templates in all keys, cf. https://github.com/ufs-community/uwtools/blob/main/src/uwtools/config/jinja2.py#L134

Author: Whoa! Neat! I didn't even realize.
